File: discord_rpc_plugin.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Discord RPC Plugin
مصمم لعرض حالة مخصصة على الديسكورد
"""

import time
import json
import os
from pypresence import Presence

class DiscordRPCPlugin:
    def __init__(self, config_file='config.json'):
        """تهيئة البلوجين"""
        self.config = self.load_config(config_file)
        self.client_id = self.config.get('application_id')
        self.rpc = None
        
    def load_config(self, config_file):
        """تحميل الإعدادات من ملف JSON"""
        if os.path.exists(config_file):
            with open(config_file, 'r', encoding='utf-8') as f:
                return json.load(f)
        else:
            print(f"⚠️ ملف الإعدادات {config_file} غير موجود")
            return self.default_config()
    
    def default_config(self):
        """الإعدادات الافتراضية"""
        return {
            "application_id": "YOUR_APP_ID_HERE",
            "state": "Fabric 1.21.11",
            "details": "Playing Minecraft",
            "large_image": "minecraft",
            "large_text": "SKLauncher 4.0",
            "small_image": "online",
            "small_text": "Online",
            "update_interval": 15
        }
    
    def connect(self):
        """الاتصال بـ Discord"""
        try:
            self.rpc = Presence(self.client_id)
            self.rpc.connect()
            print(f"✅ متصل بـ Discord بنجاح!")
            return True
        except Exception as e:
            print(f"❌ خطأ في الاتصال: {e}")
            return False
    
    def update_presence(self, state=None, details=None, large_image=None, 
                       large_text=None, small_image=None, small_text=None):
        """تحديث حالة الديسكورد"""
        if not self.rpc:
            print("❌ لم تتصل بـ Discord بعد")
            return False
        
        try:
            self.rpc.update(
                state=state or self.config.get('state'),
                details=details or self.config.get('details'),
                large_image=large_image or self.config.get('large_image'),
                large_text=large_text or self.config.get('large_text'),
                small_image=small_image or self.config.get('small_image'),
                small_text=small_text or self.config.get('small_text'),
                start=int(time.time())
            )
            print(f"✅ تم التحديث: {details}")
            return True
        except Exception as e:
            print(f"❌ خطأ في التحديث: {e}")
            return False
    
    def set_custom_status(self, app_name, details, state, large_image=None, small_image=None):
        """تعيين حالة مخصصة"""
        config = {
            "state": state,
            "details": details,
            "large_image": large_image or "default",
            "large_text": app_name,
            "small_image": small_image or "online",
            "small_text": "Online"
        }
        self.update_presence(**config)
    
    def keep_alive(self):
        """إبقاء الاتصال نشطاً"""
        try:
            while True:
                time.sleep(self.config.get('update_interval', 15))
                self.update_presence()
        except KeyboardInterrupt:
            print("\n⏹️ إيقاف البلوجين...")
            self.disconnect()
    
    def disconnect(self):
        """قطع الاتصال بـ Discord"""
        if self.rpc:
            self.rpc.close()
            print("🔌 تم قطع الاتصال")


def main():
    """الدالة الرئيسية"""
    print("🎮 Discord RPC Plugin")
    print("=" * 40)
    
    # إنشاء instance من البلوجين
    plugin = DiscordRPCPlugin()
    
    # الاتصال
    if plugin.connect():
        # تحديث الحالة
        plugin.update_presence()
        
        # إبقاء الاتصال نشطاً
        try:
            plugin.keep_alive()
        except KeyboardInterrupt:
            plugin.disconnect()
    else:
        print("❌ فشل الاتصال بـ Discord")
        print("تأكد من تثبيت pypresence:")
        print("  pip install pypresence")


if __name__ == "__main__":
    main()

File: discord-rpc-plugin.js
#!/usr/bin/env node
/**
 * Discord RPC Plugin (Node.js)
 * مصمم لعرض حالة مخصصة على الديسكورد
 */

const { Client } = require('discord-rpc');
const fs = require('fs');
const path = require('path');

class DiscordRPCPlugin {
    constructor(configFile = 'config.json') {
        this.config = this.loadConfig(configFile);
        this.clientId = this.config.application_id;
        this.client = null;
        this.updateInterval = this.config.update_interval || 15;
    }

    loadConfig(configFile) {
        try {
            if (fs.existsSync(configFile)) {
                const data = fs.readFileSync(configFile, 'utf-8');
                return JSON.parse(data);
            } else {
                console.warn(`⚠️ ملف الإعدادات ${configFile} غير موجود`);
                return this.defaultConfig();
            }
        } catch (error) {
            console.error(`❌ خطأ في قراءة الإعدادات: ${error.message}`);
            return this.defaultConfig();
        }
    }

    defaultConfig() {
        return {
            application_id: 'YOUR_APP_ID_HERE',
            state: 'Fabric 1.21.11',
            details: 'Playing Minecraft',
            large_image: 'minecraft',
            large_text: 'SKLauncher 4.0',
            small_image: 'online',
            small_text: 'Online',
            update_interval: 15
        };
    }

    async connect() {
        try {
            this.client = new Client({ transport: 'ipc' });
            
            this.client.on('ready', () => {
                console.log('✅ متصل بـ Discord بنجاح!');
                this.updatePresence();
            });

            await this.client.login({ clientId: this.clientId });
            return true;
        } catch (error) {
            console.error(`❌ خطأ في الاتصال: ${error.message}`);
            return false;
        }
    }

    async updatePresence(state = null, details = null, largeImage = null, 
                         largeText = null, smallImage = null, smallText = null) {
        if (!this.client) {
            console.error('❌ لم تتصل بـ Discord بعد');
            return false;
        }

        try {
            await this.client.setActivity({
                state: state || this.config.state,
                details: details || this.config.details,
                largeImageKey: largeImage || this.config.large_image,
                largeImageText: largeText || this.config.large_text,
                smallImageKey: smallImage || this.config.small_image,
                smallImageText: smallText || this.config.small_text,
                startTimestamp: new Date()
            });
            console.log(`✅ تم التحديث: ${details || this.config.details}`);
            return true;
        } catch (error) {
            console.error(`❌ خطأ في التحديث: ${error.message}`);
            return false;
        }
    }

    async setCustomStatus(appName, details, state, largeImage = null, smallImage = null) {
        await this.updatePresence(
            state,
            details,
            largeImage || 'default',
            appName,
            smallImage || 'online',
            'Online'
        );
    }

    async keepAlive() {
        console.log(`⏱️  سيتم التحديث كل ${this.updateInterval} ثانية...`);
        setInterval(() => {
            this.updatePresence();
        }, this.updateInterval * 1000);
    }

    async disconnect() {
        if (this.client) {
            await this.client.destroy();
            console.log('🔌 تم قطع الاتصال');
        }
    }
}

// البرنامج الرئيسي
async function main() {
    console.log('🎮 Discord RPC Plugin (Node.js)');
    console.log('='.repeat(40));

    const plugin = new DiscordRPCPlugin();

    if (await plugin.connect()) {
        await plugin.keepAlive();

        // التعامل مع إيقاف البرنامج
        process.on('SIGINT', async () => {
            console.log('\n⏹️ إيقاف البلوجين...');
            await plugin.disconnect();
            process.exit(0);
        });
    } else {
        console.error('❌ فشل الاتصال بـ Discord');
        console.error('تأكد من تثبيت discord-rpc:');
        console.error('  npm install discord-rpc');
        process.exit(1);
    }
}

// تصدير الكلاس للاستخدام كـ module
module.exports = DiscordRPCPlugin;

// تشغيل البرنامج إذا تم استدعاؤه مباشرة
if (require.main === module) {
    main().catch(console.error);
}
